cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1024 @@
+                             std::unique_ptr<ShredNode>* node) {
+#define CASE(data_type) \
+    case data_type::type_id: \
+      return PrimitiveShredNode<data_type>::Create(field, rep_level, def_level, pool, node);
+  switch (field->type()->id()) {
+    case Type::STRUCT:
+      return StructShredNode::Create(field, rep_level, def_level, pool, node);
+    case Type::LIST:
+      return ListShredNode::Create(field, rep_level, def_level, pool, node);
+    PRIMITIVE_CASE_LIST
+    default:
+      return Status::Invalid("Unsupported type: ", field->type()->name());
+  }
+#undef CASE
+}
+
+} // anonymous namespace
+
+class Shredder::Impl {
+ public:
+  Impl(std::unique_ptr<ShredNode> root) : root_(std::move(root)) {}
+
+  std::unique_ptr<ShredNode> root_;
+};
+
+Status Shredder::Create(std::shared_ptr<Field> schema,
+                        MemoryPool* pool,
+                        std::shared_ptr<Shredder>* shredder) {
+  std::unique_ptr<ShredNode> root;
+  RETURN_NOT_OK(ShredNode::CreateTree(schema, 0, 0, pool, &root));
+
+  std::unique_ptr<Impl> impl(new Impl(std::move(root)));
+
+  shredder->reset(new Shredder(std::move(impl)));
+  return Status::OK();
+}
+
+Shredder::Shredder(std::unique_ptr<Impl> impl) : impl_(std::move(impl)) {
+}
+
+Shredder::~Shredder() {
+}
+
+Status Shredder::Shred(const Array& array) {
+  if (!array.type()->Equals(impl_->root_->field()->type())) {
+    return Status::Invalid("Array schema doesn't match shredder schema");
+  }
+  for (int64_t i = 0; i < array.length(); i++) {
+    RETURN_NOT_OK(impl_->root_->Handle(array, i, 0));
+  }
+  return Status::OK();
+}
+
+Status Shredder::Shred(const ChunkedArray& array) {
+  for (int i = 0; i < array.num_chunks(); i++) {
+    RETURN_NOT_OK(Shred(*array.chunk(i)));
+  }
+  return Status::OK();
+}
+
+Status Shredder::Finish(ColumnMap* colmap) {
+  return impl_->root_->FinishRecursive(colmap);
+}
+
+const std::shared_ptr<Field>& Shredder::schema() const {
+  return impl_->root_->field();
+}
+
+
+// Stitcher implementation
+namespace {
+
+// Stitching algorithm transitions through a series of states.
+//
+// Every state is given by pair of nodes (base, leaf), where
+// "leaf" is the left-most leaf of the subtree starting in "base".
+// At every state, we go down from "base" node to "leaf" node
+// and append elements on the way.
+// This process is controlled by the next definition level read from
+// def_levels array of the "leaf" node.
+// If at some point def_level instructs us to append NULL (or empty list), we stop.
+// If we reach "leaf" node, we consume and append value.
+//
+// Next we look at the next repetition level in rep_levels array of the leaf node
+// and use prebuilt FSM tables to perform transition to the next state (base1, leaf1).
+//
+// The process is terminated when level arrays become exhausted.
+//
+// Below example demonstrates schema and possible sequence of states.
+// Nodes with [] are lists, other non-leaf nodes are structs.
```

Comment:
How might you handle unions?

Reply:
I do not support unions as of now. When (and if) unions get support, I will update this example with unions.